### DISCUSSION TITLE
Include wider taxa in the Help Identify UI

### DISCUSSION THREAD

**Discussion by nimmolo** [MushroomObserver/mushroom-observer#1404] [Ideas]
With the "Help Identify" UI, as it’s presently set up, once an observation gets a proposed name other than "Fungi" with any confidence above “doubtful", it will be removed from “Help Identify” for all users. Only the id “Fungi” shows up regardless of confidence. 

This was brought to my attention by `Excited_delirium`, who has been tagging a lot of unidentifiables as "Agaricales sensu lato". Narrowing down the taxon to such a wide taxon, while it may be a step in the right direction, has two effects:
- _Takes it off the radar_ for “help identify” users who may recognize it because they are specialists for some taxon - this is probably not helpful unfortunately
- _Puts it on the radar_ for people who search for “Agaricales sensu lato” — but they’d have to search for that in the Observations index, not here

Even though this is not particularly helpful, I'm not sure we will have much success discouraging people from doing what Excited Delirium's doing, because in one sense, it's a pretty common-sense start on identification. 

The situation instead suggests to me that there could be some alterations to “Help identify”.  I feel like we could use an interface to filter which “wider” taxa to include in the dragnet.

I'd like to get some suggestions for those. I guess a db check would be useful, but maybe it should follow the ridgeline of a taxonomic hierarchy instead?

**Comment by nimmolo** [comment]
@mo-nathan has suggested a [different approach](https://github.com/MushroomObserver/mushroom-observer/pull/1426#issuecomment-1468035633) to the one i'm currently working with, for redefining observations needing IDs. 

Below is a summary of the current state, the working PR, and Nathan's suggestions:

The current "Help Identify" interface shows the results of the following complex runtime query:

> (Observations where the ID is "Fungi", or where the vote cache < 0 (unconfident)) where the current user has not already voted on a naming, and not signaled they are not interested.

This kind of assumes that users have taken our suggestion, and put "Fungi" as the name, when they weren't sure. However, as noted above, other users are going around and refining the naming to something a bit more specific, yet still very general. Individual users may "mark an observation as reviewed" for their own session, but we don't have a way for an experienced user to take it off the radar of all users. The query is slow-ish, it hits Observations, Namings, Votes, and ObservationViews.

> side note: Nathan brought up the debatability of "confident naming", but I don't feel there's any need to reinvent it. It's defined already on MO as `vote_cache <= 0`, with vote weight distributed unequally by experience. That threshold could be tweaked, but I'm fine with it. 

My new PR #1426 is proposing to widen the initial scope like this:

> (Observations whose naming is above Genus, or with more specific unconfident namings), where the current user has not already voted on a naming, and not signaled they are not interested.

In other words, the new PR is asserting that any ID above Genus "needs further identification", even if that may not be possible. It's also a more complex and slower query. 

The nice thing about this query is that further filtering of the same query allows, say, an expert in _Inocybaceae_ to view observations without specific ID that are nonetheless considered to be under the family of _Inocybaceae_. Such a specialist could then review results. 

Likewise with locations: a person familiar with Wisconsin mushrooms could filter for Wisconsin observations with generic or above IDs, and help identify. 

But each one of these filters performs the complex initial query again, with further conditions.

I believe Nathan's saying, stop the madness: the query would be a lot simpler if we cached the results of this complex query for all existing observations, into a new boolean column `Observation.needs_id`. We'd rebuild it only on new observations/namings going forward, and we'd allow experienced users direct access to alter the column - they could mark it "reviewed" and it would stop showing up for everybody. Then, for the "Help Identify" index, we could just query the results of this column in the Observations table.

A further possibility is to let the observer themselves flag the observation as "needing ID", and clear the flag as soon as a naming gets a certain threshold of confidence.

**Reply by nimmolo** [reply]
One more note: I feel like this taxon-and-below filter I'm writing, "Observations of a higher taxon or of any of its subtaxa", could be a valuable new kind of query that "Help Identify" users would use. It could also be nice on the main obs index.

That query might be too slow to implement on obs index (?) and it would probably be slowish even with the above change for Help Identify. Still (for Help Identify), preselecting for `observation.needs_id` would narrow the initial scope of that query quite a bit.

**Comment by pellaea** [comment]
Great summary, thanks, Nimmo!  I'll try to give my feedback later this
evening if time permits, but it sounds like y'all are on the right track
with some good ideas.

**Comment by mo-nathan** [comment]
Thanks for writing this up!  It might helpful to step back and think about this problem in terms of requirements rather than detailed implementation.  In my view, the goal of this feature is to prioritize the review of observations that our community members can provide better ids for.  Observations that have no proposed names above 0 are pretty obviously at the top of this list.  All remaining observations are somewhere on this list with the priority being some combination of the existing votes, the taxonomic level, and changes in relevant taxonomy.  There is also the practical question of where to cut off the list.  I also think we want a mechanism for bumping an observation up the list.

**Reply by nimmolo** [reply]
@mo-nathan - Agree about the desirability of being able to bump an obs up the list.

In case you missed it initially, the first post in this discussion is where i brought up some ways to improve upon the current cutoff at "Fungi", relevant to the goal you stated above (which I agree with).

The main thing i'd like to deal with is that nonspecific namings below "Fungi" are widespread... and currently at least, counterproductive, because they remove the obs from "Help Identify". 

Nonspecific ids can potentially expedite community identification by getting the obs to the attention of
- specialists of the higher taxon
- regional experts 
- collaborators (user filter)
Someone like Alan can say "idk, but it seems like Rhodophyllaceae" and another user specialized in the family could winnow it down a bit. But currently one cannot help identify obs with, say a Family-level naming, via "Help identify".

However, it's already possible to get a taxonomic filter by using "Advanced search" on the observation index, rather than the "Help Identify" index. Maybe what we need is just a UI change on help identify, so that if people want to use a taxonomic filter, it takes them to the regular observation index.

**Comment by mo-nathan** [comment]
Yeah, I think we're pretty much on the same page.  I wasn't disagreeing.  I
was just trying to write down what the goals are.  I'm delighted you've
taken this on.  It's one of those ideas that's been rattling around in my
head for years.

I'm not sure if this is what you're already suggesting, but it seems like
having the filter on the "Help Identify" might be helpful.  The reality is
that I don't know how many users we have that are specialized on particular
taxa.  Of course we have folks like Rod Tulloss with Amanitaceae, but I
doubt we have anyone focused on the Rodophyllaceae, but maybe.  I guess all
I'm saying is that before you sink time into it, you might want to figure
out who might use it.  I also think the more configuration it requires the
fewer users will figure out how to use it.  Ideally we'd do something like
look at the taxa that they have previously identified and use that to
filter out what they might have opinions on.  Again we'd want to find out
who might use such a feature or how many hits they'd really get, but I
could certainly imagine it being pretty sweet to have a page that I went to
that showed me the stuff that it thought I would be particularly good at
identifying.

**Reply by nimmolo** [reply]
Thanks @mo-nathan — yes, i was thinking to write the filter for "Help Identify", and then i guess i was wondering again if it should all be on the Observations index.

There's a practical UX consideration that keeps the Help Identify index from potentially being a simple query of the Observations table, even if we cache the results in a new column — we're accounting for the session user's possible existing vote and "mark as reviewed". That's making us check two other tables at runtime (votes and obs_views).

I wasn't picturing trying to guess what the user would be interested in identifying, via their history though, that does sound crazy complicated. I was just going to leave it to them to manually filter via a text field (a term) and a select dropdown (type of query: taxon, region or user condition) in the search bar. I do have something working. The way I've written it, these filters add further conditions to the above query (and the ui doesn't allow more than one of these at once).  I think people would use this, even at my limited state of knowledge I only feel comfortable in a few taxa, or in a specific region.

It is certainly a complex runtime query though, as you aptly put it. I just can't imagine how it could not be! But i'm all ears.

**Comment by mo-nathan** [comment]
For now doing it the way it sounds like you've done it is fine.  We can
worry about the caching bit if it gets too slow or it becomes clear that we
need to provide an easy way to override the query.

**Reply by nimmolo** [reply]
Thanks Nathan. 

I have another question: One thing Jason and I tried when we initially set this up was to make an attempt to inherit the user's previous query... that is, if they'd been searching for `Amanita` in `Illinois, USA`, then we'd try to inherit that filter on needs ID. That's not really working, though, for reasons that I guess are obvious now. There are quite a few filters a user could apply to the regular observation index, that would cause the `needs_id` query to return zero results every time, and this may not be intuitively obvious to the user. 

Furthermore, `needs_id` must create its own query,  for the sake of paginating results. So i'm thinking that I need to think of this page as always resetting the query, and applying/saving its own filters to that query. These filters will apply only when the user is trying to "Help Identify", and will not be usable after leaving the Needs Identification page. 

Does that make sense?

**Comment by nimmolo** [comment]
Just realized, by reading through the class `ContentFilter`, that we have cached columns on `Observations` for `classification` and `where`, so I don't have to do the separate query on `Names` or `Locations` to figure out a clade or region for an observation. I can't tell if it's speeding up the `needs_id` query, but it seems like it should.

**Reply by nimmolo** [reply]
Rewrote to search those columns instead of doing joins.
